doc/developer/design/README.md

```diff
@@ -35,54 +37,54 @@ Examples (some "in spirit", because the design doc template changed over time):
-writer. The appropriate level of technical detail depends on the target
-reviewers and the nature of the design/change that is being proposed.
-
-Ask your tech lead if you are unsure about the right level of technical detail.
-A good rule of thumb is that you should strive for the *minimum* level of
-detail that fully communicates the proposal to your reviewers. Going into too
-much detail has a cost: it takes longer to write the document and longer to
-review the document, plus the finest details are those that are most likely to
-change during implementation.
+As stated earlier, the goal of a design document is to thoroughly
+discover problems and examine potential solutions before moving
+into the delivery phase of a project. In order to do this, each
+design document should:
+
+- Clearly and succinctly state the problem it aims to solve,
+  along with any required context.
+- Document evidence of the problem via customer interviews,
+  metrics, or any other means.
+- Crisply articulate the author’s preferred solution to the
+  problem, and why.
+- Document all alternative solutions that were considered,
+  and why they weren’t chosen.
+- Document any dependencies that may need to break or change
+  as a result of this work.
+- Record any decisions that resulted from the document.
+
+If any of these goals are unclear to you, or if the specific
+requirements of the document are unclear once you kick off a
+design process, reach out to your manager.
 
 ## Non-Goals
 
-These are not explicitly required for a good design document. You should,
-however, use your own judgement to determine what is and isn't required for
-your specific case.
+These are not explicitly required for a good design document.
+You should, however, use your own judgment to determine what
+is and isn’t required for your specific case.
 
- - Allow a drive-by commenter without any context to fully understand the
-   design: the intention is that a group of people with common context can come
-   together, understand the design and come to a decision. Some context might
-   be gathered from previous design documents, from reference documentation, or
-   from looking at the code.
- - Provide a fully spec'ed reference implementation: depending on the topic, it
-   can be very helpful to prepare a prototype. Both for validating the design
-   and for presenting it to others. However, it is not always a required part
-   of the design.
+Allow a drive-by commenter without any context to fully understand
+the design. The intention is that a group of people with some
+common context can come together, understand the design, and
+come to a decision. Provide a fully spec’ed out implementation
+of any of the solutions. Depending on the topic, it can be very
+helpful to provide a prototype, both for validating the design
+and for presenting it to others. However, it is not a required
+part of the design.
 
 ## When should you make a design document?
```

Comment:
> Does it make sense to note down here that the design document should be written and completed before starting to work on the code (so that design doc changes resulting from a discussion don't require reworking the code)?

I wouldn't say "completed" but "reviewed" or "discussed" in that case.In my mind a design doc should not be considered "completed" until the code that implements it is written.

Asking to have a design doc up until the point where all unknowns are resolved before we can start with the implementation means failing to acknowledge that for almost all design docs we are going to encounter unknown unknowns when we start working on the coding part.

We can (and probably should) merge the doc after the initial round of reviews, but we should make sure that the content of the design doc is aligned with the code that realizes it.

Reply:
I agree with Alexander on this one. Speaking personally, the design docs I've worked on would be much lower quality if we hadn't done some implementation work first. (And given the process here, "fixing" a design doc is pretty expensive... it needs to go through the same set of reviews as the original doc. That means it's better to spend a little time to get it right up front!)

There's definitely _some_ line, though. Maybe it's fair to say that the design docs should land before the proposed feature becomes customer-visible? So prototyping on a branch or behind a flag would be fine.

Reply:
Agree! Added a note on prototypes/experimental work in 18e5a1148b.